source/ncengine/graphics2/frontend/subsystem/InstanceCache.h

```diff
@@ -0,0 +1,198 @@
+// from the graphics backend and game logic, respectively.
+class InstanceCache
+{
+    public:
+        explicit InstanceCache(uint32_t maxMeshRenderers,
+                               uint32_t initialBatchSize = 1u);
+
+        // Get the staging area for enqueueing buffer writes.
+        auto GetStagingArea() -> InstanceCacheStaging&
+        {
+            return m_stagingArea;
+        }
+
+        // Build info specifying modified buffer ranges.
+        auto BuildState() -> BufferUpdateInfo<InstanceData>;
+
+        // Generate draw call information for batches. The return value contains a collection of batches
+        // to be drawn per material pass provided in the input 'passes'.
+        auto BuildBatches(std::span<const MaterialPass::type> passes) -> std::vector<std::vector<Batch>>;
+
+        // Merge staging area into the buffer.
+        // IMPORTANT: Staging area cannot be modified while this is running (game logic cannot be running).
+        void CommitPendingChanges();
+
+        // Clean and compress unused state. Intended for scene changes but could also be called periodically, if needed.
+        // - Removes all empty batches
+        // - Shrinks excess batch capacity
+        // - Compacts buffer by shifting instances/batches towards the front
+        // IMPORTANT: Staging area cannot be modified while this is running (game logic cannot be running).
+        void Purge() noexcept;
+
+        auto IsValidInstance(uint32_t entityId) const -> bool
+        {
+            return m_entityToIndex.contains(entityId);
+        }
+
+        auto GetInstanceIndex(uint32_t entityId) const -> uint32_t
+        {
+            return m_entityToIndex.at(entityId);
+        }
+
+        auto GetEntityId(uint32_t instanceIndex) const -> uint32_t
+        {
+            return m_indexToEntity.at(instanceIndex);
+        }
+
+        auto GetInstanceData(uint32_t instanceIndex) const -> const InstanceData&
+        {
+            return m_buffer.at(instanceIndex);
+        }
+
+        auto HasBatchFor(const BatchKey& key) -> bool
+        {
+            return std::ranges::contains(m_batches, key, &BatchRegion::key);
+        }
+
+        auto GetRegion(const BatchKey& key) -> BatchRegion&
+        {
+            auto pos = std::ranges::find(m_batches, key, &BatchRegion::key);
+            NC_ASSERT(pos != m_batches.end(), "Batch not found");
+            return *pos;
+        }
+
+        auto GetInstanceCountUpperBound() const -> size_t
+        {
+            return m_buffer.size();
+        }
+
+        auto GetBatchCount() const -> size_t
+        {
+            return m_batches.size();
+        }
+
+    private:
+        std::vector<InstanceData> m_buffer;
+        std::vector<BatchRegion> m_batches;
+        sparse_map<uint32_t> m_entityToIndex;
+        std::vector<uint32_t> m_indexToEntity;
+        InstanceCacheStaging m_stagingArea;
+        uint32_t m_dirtyBegin = UINT_MAX;
```

Comment:
What did we decide on UINT_MAX vs std::numeric_limits<uint32_t>::max()? Obviously one is fewer characters, so I'm fine with it. I'm actually curious why both exist.

Reply:
Fixed. I usually use `UINT_MAX` when POCing things and fix later. There's a small emotional burden to breaking flow state to include `<limits>` that I just can't afford :)